lib/auth/methods.go

```diff
@@ -110,41 +114,50 @@ func (s *Server) AuthenticateUser(req AuthenticateUserRequest) error {
 	return err
 }
 
-func (s *Server) authenticateUser(ctx context.Context, req AuthenticateUserRequest) error {
+func (s *Server) authenticateUser(ctx context.Context, req AuthenticateUserRequest) (*types.MFADevice, error) {
 	if err := req.CheckAndSetDefaults(); err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
 	authPreference, err := s.GetAuthPreference()
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
 	switch {
 	case req.U2F != nil:
 		// authenticate using U2F - code checks challenge response
 		// signed by U2F device of the user
+		var mfaDev *types.MFADevice
 		err := s.WithUserLock(req.Username, func() error {
-			return s.CheckU2FSignResponse(ctx, req.Username, &req.U2F.SignResponse)
+			var err error
+			mfaDev, err = s.CheckU2FSignResponse(ctx, req.Username, &req.U2F.SignResponse)
+			return err
 		})
 		if err != nil {
 			// provide obscure message on purpose, while logging the real
 			// error server side
 			log.Debugf("Failed to authenticate: %v.", err)
-			return trace.AccessDenied("invalid U2F response")
+			return nil, trace.AccessDenied("invalid U2F response")
 		}
-		return nil
```

Comment:
Nit: I would add `if mfaDev == nil { return nil, trace.XXX }` check here and below, just in case.